server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/StringTermsAggregatorFromFilters.java

```diff
@@ -0,0 +1,230 @@
+import org.elasticsearch.search.aggregations.support.AggregationContext;
+import org.elasticsearch.search.aggregations.support.ValuesSourceConfig;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.function.LongPredicate;
+
+import static org.elasticsearch.search.aggregations.InternalOrder.isKeyOrder;
+
+/**
+ * Adapts a {@code terms} aggregation into a {@code filters} aggregation.
+ * Its possible that we could lovingly hand craft an {@link Aggregator} that
+ * uses similar tricks that'd have lower memory overhead but generally we reach
+ * for this when we know that there aren't many distinct values so it doesn't
+ * have particularly high memory cost anyway. And we expect to be able to
+ * further optimize the {@code filters} aggregation in more cases than a
+ * special purpose {@code terms} aggregator.
+ */
+public class StringTermsAggregatorFromFilters extends AdaptingAggregator {
+    static StringTermsAggregatorFromFilters adaptIntoFiltersOrNull(
+        String name,
+        AggregatorFactories factories,
+        AggregationContext context,
+        Aggregator parent,
+        boolean showTermDocCountError,
+        CardinalityUpperBound cardinality,
+        Map<String, Object> metadata,
+        ValuesSourceConfig valuesSourceConfig,
+        BucketOrder order,
+        BucketCountThresholds bucketCountThresholds,
+        LongPredicate acceptedOrds,
+        SortedSetDocValues values
+    ) throws IOException {
+        if (false == valuesSourceConfig.alignesWithSearchIndex()) {
+            return null;
+        }
+        List<String> keys = new ArrayList<>();
+        List<Query> filters = new ArrayList<>();
+        TermsEnum terms = values.termsEnum();
+        for (long ord = 0; ord < values.getValueCount(); ord++) {
+            if (acceptedOrds.test(ord) == false) {
+                continue;
+            }
+            terms.seekExact(ord);
+            keys.add(Long.toString(ord));
+            /*
+             * It *feels* like there should be a query that operates
+             * directly on the global ordinals but there isn't. Building
+             * one would be tricky because there isn't mapping from
+             * every global ordinal to its segment ordinal - only from
+             * the segment ordinal to the global ordinal. You could
+             * search the mapping to get it but, like I said, tricky.
+             */
+            TermQueryBuilder b = new TermQueryBuilder(
+                valuesSourceConfig.fieldContext().field(),
+                valuesSourceConfig.format().format(terms.term())
+            );
+            filters.add(context.buildQuery(b));
+        }
+        StringTermsAggregatorFromFilters adapted = new StringTermsAggregatorFromFilters(
+            parent,
+            factories,
+            subAggs -> FiltersAggregator.buildFilterOrderOrNull(
```

Comment:
I see why you did that hack with null earlier. I think we can extract first 3 conditions from `buildFilterOrderOrNull` into something like `canBuildFiter`. Then based on the result of this check you can decide if you want to try building adapted or not.

Reply:
I did something pretty similar when I did this for ranges and wasn't totally happy with it so I wanted to try this way. I'll take another look.

Reply:
I would be happy to iterate on it tomorrow. It's just doesn't feel right to produce an intentionally broken object even before discarding it immediately upon creation if we can easily prevent creation of this object in the first place. Maybe we can remove this convoluted factory logic from the AdaptingAggregator constructor all together. It's used in 3 places and it really messes things up in one of them, maybe it doesn't really belong there.

Reply:
I did make that method and I like this better. Check out `RangeAggregator.adaptIntoFiltersOrNull` for a method that has a better excuse to be shaped the way it is.

Reply:
The factory logic is genuinely weird. I'd love to be able to drop it but it'd make the `parent` wonky. I don't recall off hand exactly what weird thing made me do it. But it was weird.